src/main.rs

```diff
@@ -486,17 +486,173 @@ fn shell_exec(vm: &VirtualMachine, source: &str, scope: Scope) -> ShellExecResul
     }
 }
 
+struct ShellHelper<'a> {
+    vm: &'a VirtualMachine,
+    scope: Scope,
+}
+
+impl ShellHelper<'_> {
+    fn complete_opt(&self, line: &str) -> Option<(usize, Vec<String>)> {
```

Comment:
I would suggest to reduce the size of this function. Try to create functions that can be displayed without having to scroll.